[SOL-1954, SOL-1955, SOL-1956] New Receipt Page - Frontend

marjev opened the pull request:
@mjfrey Here's a frontend part of the Receipt Page PR. I don't want to split this any further because I'm afraid we might lose the bigger picture that this PR offers. Any suggested changes should be applied, if needed, across the entire PR.

@vkaracic please review as well.

Acceptance tests will be the 3rd PR.

- [x] UX review
- [x] Text review

Latest screenshot:

![image](https://cloud.githubusercontent.com/assets/6833568/20920160/2a3b5fb6-bb9f-11e6-9eab-4d6680e5b0b3.png)

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=978&repo=edx%2Fecommerce

mjfrey commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,161 @@ def get_context_data(self, **kwargs):
+
+            seat = order.lines.first().product
+            site_configuration = self.request.site.siteconfiguration
+            course = site_configuration.course_catalog_api_client.course_runs(seat.attr.course_key).get()
```
actually -- why don't we just use the course info already in otto.  No need to call the catalog service.

clintonb replied:
I have a similar question. I see no usage of this data aside from `course.key`, which you already have.

marjev replied:
Good catch. This can be removed entirely. This was added because the Old Receipt Page design had Course name on the page.

I'll remove this.

mjfrey commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,161 @@ def get_context_data(self, **kwargs):
+
+            try:
+                processor_response = PaymentProcessorResponse.objects.filter(basket=order.basket).latest()
+                if processor_response and processor_response.processor_name == 'paypal':
```
This logic should be part of the ProcessorResponse class as a method.  get_recipient_info.

mjfrey commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,161 @@ def get_context_data(self, **kwargs):
+            'for_help_text': _(
+                "If your course does not appear on your dashboard, contact {payment_support_link}."
+            ).format(payment_support_link=payment_support_link),
+            'is_payment_complete': True,
```
is this ever False?

marjev replied:
yes, it is, when the payment is not complete, which is some use-case with CyberSource I'm not too familiar with and didn't have time to go into (would take me additional days just to figure that one out).

We'll have to figure out how we want to handle CyberSource use-case in our code. Clinton suggested what seems to be a solid solution to this problem. I'll try to implement it that way.

mjfrey commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,161 @@ def get_context_data(self, **kwargs):
+            'is_payment_complete': True,
+            'lms_url': get_lms_url(),
+            'name': '{} {}'.format(self.request.user.first_name, self.request.user.last_name),
+            'nav_hidden': True,
```
same here

mjfrey commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -122,3 +145,112 @@ def test_post_returns_payment_support_email_in_context(self):
+        httpretty.register_uri(httpretty.POST, self.path, body=post_data)
+
+        response = self.client.post(self.path, params={'basket_id': 1}, data=post_data)
+        self.assertEqual(response.status_code, 200)
```
this is a helper function not a test.  You should check for this in the actual test that calls it.

mjfrey commented:
@clintonb can you take a look?

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -1,25 +1,28 @@
+from ecommerce.extensions.api.serializers import OrderSerializer
 from ecommerce.extensions.checkout.exceptions import BasketNotFreeError
 from ecommerce.extensions.checkout.mixins import EdxOrderPlacementMixin
+from ecommerce.extensions.checkout.utils import add_currency, get_credit_provider_details, get_receipt_page_url
```
`add_currency` has a couple issues:

1. The name is confusing. `format_price` is less confusing.
2. It defaults to the currency from settings. While we may only use a single currency today, that will not always be the case. Correcting this to accept both a currency and an amount will prevent a lot of headaches in the future.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+                else:
+                    recipient_info = None
+            except (PaymentProcessorResponse.DoesNotExist, KeyError):
+                recipient_info = None
```
If you were to keep this code (which I recommend against, since you should not be using `PaymentProcessorResponse`), you should log the exception alongside a helpful message. Otherwise, debugging will be painful.

marjev replied:
I decided to do what you recommended.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+                'lms_url': order.site.siteconfiguration.lms_url_root,
+                'provider_data': provider_data,
+                'receipt': receipt,
+                'verified': seat.attr.certificate_type == 'verified'
```
This is redundant given that you have `is_verification_required`.

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -122,3 +145,112 @@ def test_post_returns_payment_support_email_in_context(self):
+        }
+
+        self.assertEqual(response.status_code, 200)
+        self.assertDictContainsSubset(receipt_data, response.context_data['receipt'])
```
Add the data to `context_data` and remove this assertion.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
         return context
+
+
+class ReceiptResponseView(TemplateView):
```
This view needs to validate access to the requested order. Orders should be accessible if either of the following conditions is met:

1. The user is a staff user.
2. The user placed the order (e.g. `order.user == request.user`)

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+
+        site_configuration = self.request.site.siteconfiguration
+        payment_support_email = site_configuration.payment_support_email
+        payment_support_link = '<a href="mailto:{email}">{email}</a>'.format(email=payment_support_email)
```
No HTML in views. Return the email address only. The template should handle the rendering. Also, this should probably be part of the core context processor since it already contains the support URL.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+        context.update({
+            'dashboard': site_configuration.build_lms_url('/dashboard'),
+            'error_summary': _('An error occurred while creating your receipt.'),
+            'error_text': None,
```
Is this necessary? If this field is not present, will the page still render?

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+            'dashboard': site_configuration.build_lms_url('/dashboard'),
+            'error_summary': _('An error occurred while creating your receipt.'),
+            'error_text': None,
+            'for_help_text': _(
```
Why isn't this in the template?

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -110,3 +114,162 @@ def get_context_data(self, **kwargs):
+            'page_title': _('Receipt'),
+            'payment_support_email': payment_support_email,
+            'payment_support_link': payment_support_link,
+            'platform_name': settings.SITE_NAME,
```
This is part of the core context processor.

clintonb commented on ecommerce/static/js/routers/receipt_router.js:
```diff
@@ -0,0 +1,24 @@
+
+            showReceiptPage: function(orderNumber) {
+                var page = new ReceiptPage({orderNumber: orderNumber});
+                this.currentView = page;
```
Merge with the line above. No need for the variable if it's not being used.

clintonb commented on ecommerce/templates/oscar/checkout/receipt.html:
```diff
@@ -0,0 +1,224 @@
+                  <table class="table">
+                    <thead class="order-columns">
+                      <tr>
+                        <th scope="col" >{% trans "Qty" %}.</th> <!-- Translators: Abbreviation for quantity.-->
```
This comment should be above the line marked for translation.

clintonb commented on ecommerce/templates/oscar/checkout/receipt.html:
```diff
@@ -0,0 +1,224 @@
+                               </div>
+                           {% endfor %}
+                            </td>
+                            {% if receipt.discount %}
```
Move the if-statement inside the div, and de-dupe.

clintonb commented on ecommerce/static/js/apps/receipt_app.js:
```diff
@@ -0,0 +1,14 @@
+require([
```
The app and router seem unnecessary for a single page.

marjev replied:
App simply initializes the router, which initializes the page and passes the Order number to it. Receipt Backbone View uses the Order number to trigger Analytics events that are a requirement for [SOL-1956](https://openedx.atlassian.net/browse/SOL-1956).

If you know a way to do this Django side, without using Backbone Analytics Model, please let me know.

clintonb replied:
Skip the middle man. We don't need this app file. Just load the page JS directly. Also, update `build.js`.

marjev commented on ecommerce/extensions/checkout/app.py:
```diff
@@ -7,12 +7,14 @@ class CheckoutApplication(app.CheckoutApplication):
             url(r'^free-checkout/$', self.free_checkout.as_view(), name='free-checkout'),
             url(r'^cancel-checkout/$', self.cancel_checkout.as_view(), name='cancel-checkout'),
             url(r'^error/', self.checkout_error.as_view(), name='error'),
+            url(r'^receipt/', self.receipt_response.as_view(), name='receipt'),
```
Should I remove this ```url(r'thank-you/$', self.thankyou_view.as_view(), name='thank-you'),```?

I will extend the `ThankYouView` provided by Oscar, but I'm not sure if we want to expose ThankYouView via `thank-you` URL.

clintonb replied:
Yeah, I would replace that thank-you URL with receipt.

marjev commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -122,3 +145,112 @@ def test_post_returns_payment_support_email_in_context(self):
+        post_data = {'decision': 'ACCEPT', 'reason_code': '200', 'signed_field_names': 'dummy'}
+        response = self.post_to_receipt_page(post_data)
+
+        self.assertNotContains(response, "How it Works")
```
I don't know. My guess is this was copy/pasted from the LMS.

I'll remove this.

marjev commented:
@mjfrey @clintonb Most of the comments have been addressed.

Regarding the `post` method that handles the CyberSource Payments, should I simply add a redirect to the receipt page if the CyberSource Payment didn't fail? I was thinking of using [this view](https://github.com/edx/ecommerce/commit/89ff8dc18dd97523da1132f83075c1612ef29b66#diff-3a39a1e895cafeeed9c6a0c0dffd2dceR37). Let me know what you think.

marjev commented on ecommerce/extensions/checkout/signals.py:
```diff
@@ -57,6 +58,32 @@ def track_completed_order(sender, order=None, **kwargs):  # pylint: disable=unus
     )
 
 
+@receiver(post_payment, dispatch_uid='tracking.post_payment_callback')
```
@mjfrey I decided to use `post_payment` signal as a trigger for the Segment.io event `Completed Purchase`.

Can you review this code? Do I need to send `post_payment` signal in `handle_processor_response` method for each PaymentProcessor or is that already being done somewhere in Otto/Oscar code?

marjev commented:
@clintonb While I was trying to address the 4th comment you left on the PR, I began to wonder. Since the LMS Receipt Page View has the CyberSource handling logic deeply integrated within, will removing that logic from the ecommerce Receipt Page View require additional changes to the CyberSource handling code?
I still don't know which piece of code sends the POST request to the Receipt Page View (endpoint) to begin with or at least handles to which URL the request will be made to?

FYI @mjfrey

clintonb commented on ecommerce/extensions/offer/utils.py:
```diff
@@ -60,6 +60,6 @@ def format_benefit_value(benefit):
     else:
-        converted_benefit = add_currency(Decimal(benefit.value))
-        benefit_value = _('${benefit_value}'.format(benefit_value=converted_benefit))
+        # TODO: Fix this; Currency should be a dynamic value;
```
Are you fixing this?

marjev replied:
Yes. There are tickets already created for this cleanup work.

clintonb commented on ecommerce/static/js/config.js:
```diff
@@ -12,8 +12,11 @@ require.config({
         'dataTablesBootstrap': 'bower_components/datatables/media/js/dataTables.bootstrap',
+        'date-utils': 'bower_components/date-utils/lib/date-utils.min',
         'ecommerce': 'js/shims/ecommerce',
+        'edx-ui-toolkit': '../../node_modules/edx-ui-toolkit/src/js',
```
Does this actually work? If so, we can move many of our bower-based dependencies to npm.

clintonb commented on ecommerce/static/js/pages/receipt_page.js:
```diff
@@ -0,0 +1,33 @@
+                trackPurchase(order_id);
+            }
+        },
+        trackPurchase = function(order_id) {
```
We should find a way to do this only the first time this loads. Subsequent views of this page should not fire events.

marjev replied:
Let me think about this some more.

marjev replied:
Addressed. Please take a look at 0224d38

marjev replied:
Actually, this isn't doing what I was hoping it would do, so I'll have to rethink this further, but hopefully I'll manage to send bool value from the PaymentProcessorView with the redirect.

clintonb commented on ecommerce/templates/edx/base.html:
```diff
@@ -25,6 +25,7 @@
         {% block stylesheets %}
         {% endblock %}
     {% endcompress %}
+    {% block header_extras %} {% endblock header_extras %}
```
This is not used, and should be removed.

mjfrey commented on ecommerce/extensions/checkout/utils.py:
```diff
@@ -54,18 +54,18 @@ def get_receipt_page_url(site_configuration, order_number=None):
     return format_currency(
         amount,
-        settings.OSCAR_DEFAULT_CURRENCY,
-        format=u'#,##0.00',
```
we need this format.  Why did you remove it?  This will break when the browser is in some languages.

marjev replied:
If this format is in place, the $ is not visible on the page (when the currency is set to 'USD'). Removing the format seems to fix the issue.

marjev commented on bower.json:
```diff
@@ -11,23 +11,25 @@
   ],
   "dependencies": {
     "backbone": "~1.2.1",
+    "backbone-relational": "~0.9.0",
```
Only `backbone-relational` or all of them? 
If you meant all of the dependencies, I'm worried about updating the jQuery from 1.12.2 to 3.1.1. That's a rather drastic change and perhaps should be done in a separate PR.

marjev replied:
I'll hold off from doing that because updating backbone-relational caused an unrelated test failure. I would like to keep this PR as clean and simple as possible.

clintonb commented:
Remember to incorporate these changes: https://github.com/edx/edx-platform/pull/12470

clintonb commented:
I started a Verification API at https://github.com/edx/edx-platform/commits/clintonb/verification-status-api.

marjev commented:
@clintonb @mjfrey Comments have been addressed. Vedran commented on one because we currently don't have a clear idea on how to get a basket URL for the thawed basket (in case CyberSource payment has failed). Also, I'll have to think about how to trigger the analytics event only once.

mjfrey commented on ecommerce/templates/oscar/checkout/cybersource_error.html:
```diff
@@ -0,0 +1,37 @@
+{% extends 'edx/base.html' %}
```
Do we want to call this payment_error and not use the cyber source name?

marjev replied:
If I understood correctly, this template is only for CyberSource usecase, because sometimes CyberSource Payment might "fail".

I didn't notice PayPal having this issue. There should be a subtask created on JIRA that describes this case. This template is only being called by CyberSourceInterstitialView.

clintonb commented on ecommerce/static/js/pages/receipt_page.js:
```diff
@@ -0,0 +1,34 @@
+        'use strict';
+
+        var onReady = function() {
+            var order_id = $.url('?order_number') || null,
```
`orderId`

clintonb replied:
Why bother with jQuery URL? Pass the order number as a data attribute.

clintonb commented on ecommerce/static/js/config.js:
```diff
@@ -12,8 +12,10 @@ require.config({
         'collections': 'js/collections',
         'datatables.net': 'bower_components/datatables/media/js/jquery.dataTables',
         'dataTablesBootstrap': 'bower_components/datatables/media/js/dataTables.bootstrap',
+        'date-utils': 'bower_components/date-utils/lib/date-utils.min',
```
This doesn't seem to be used anywhere.

clintonb commented on bower.json:
```diff
@@ -11,23 +11,25 @@
     "bootstrapaccessibilityplugin": "~1.0.4",
     "datatables": "1.10.10",
-    "fontawesome": "~4.3.0",
+    "date-utils": "1.2.16",
```
This doesn't seem to be used anywhere.

clintonb commented on ecommerce/static/js/pages/receipt_page.js:
```diff
@@ -0,0 +1,34 @@
+
+        return {
+            onReady: onReady,
+            trackPurchase: trackPurchase
```
This is only called by `onReady`. There is no need to expose it. If you get rid of `receipt_app.js`, there is no need to expose anything.

clintonb commented on ecommerce/templates/oscar/checkout/_credit_provider_data.html:
```diff
@@ -0,0 +1,38 @@
+      data-track-event="edx.bi.credit.clicked_complete_credit"
+      data-track-label="{{ provider.course_key }}"
+      data-track-type="click"
+      data-username="{{ name }}"
```
Why is this not username?

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -113,3 +116,92 @@ def get_context_data(self, **kwargs):
+
+    def get_context_data(self, **kwargs):
+        return {
+            'name': '{} {}'.format(self.request.user.first_name, self.request.user.last_name),
```
It seems the only place where this is used, you actually need the username of the person who placed the order. This should be `order.user.username`. If you passed the order to context, this line can go away entirely.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -113,3 +116,92 @@ def get_context_data(self, **kwargs):
+            for line in order.lines.all():
+                seat = line.product
+                if self.request.user.is_verified(self.request):
+                    if hasattr(seat.attr, 'certificate_type') and seat.attr.certificate_type == 'credit':
```
Also, this shouldn't necessarily be nested in the check for verification. Say a user has purchased a credit seat but his/her verification expired. We still want to show details for credit retrieval.

mjfrey replied:
you can only be eligible for credit if you have already verified.  Credit link will never been shown if not.

mjfrey replied:
yes I see.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -113,3 +116,92 @@ def get_context_data(self, **kwargs):
+        if self.request.user == order.user:
+            for line in order.lines.all():
+                seat = line.product
+                if self.request.user.is_verified(self.request):
```
We should only check for verification status if a seat requiring verification is in the order.

clintonb replied:
Also, @vkaracic it seems we are re-raising HTTP errors. I may have suggested this, in which case it was a bad idea. We should log the exceptions (which you are doing), but not raise them.

vkaracic replied:
Okay, I'll rework it tomorrow. Also the ``is_verified`` method accepts the request so it can construct the LMS URL.

clintonb replied:
I must have missed the addition of the `request` argument. We should avoid making our model methods dependent on `Request` objects. We should instead pass exactly the data we need (e.g. a `Site` object). This will allow them to be used outside of the scope of requests, if necessary.

clintonb replied:
@vkaracic I explored this a bit more. I can see a case where it would be useful to know the difference between the API call failing and the user not actually being verified. For example, in the event of a failure, instead of misleading the user to the verification page, we could inform the user that we were unable to check verification status and provide a link to verify in case the user is not verified. This difference is subtle; but, user experience is improved because the user is not confused if we erroneously take him/her to the verification page and he/she is already verified.

Thus, it makes sense to raise a custom error (e.g. `VerificationStatusError`) that wraps up all errors except `HttpNotFoundError`. This way callers have a single error to handle, and can choose how to inform the user.

clintonb commented on ecommerce/templates/oscar/checkout/receipt.html:
```diff
@@ -0,0 +1,148 @@
+
+              <div class="nav-wizard row">
+                {% if verification_data %}
+                  {% for key, value in verification_data.items %}
```
We only need to show one verification link. Although the URLs are course run-dependent, verification is not actually tied to the course run, just the user.

@mjfrey this needs to be updated on LMS.

clintonb replied:
Verification URLs currently follow the following format: `/verify_student/start-flow/<course-id>/`. This somewhat (although, not really) made sense when payment and verification were heavily integrated. Since the rollout of Otto, the two actions are completely separate. From a model standpoint, verifications have never been linked to courses from a model standpoint.

Ideally, verification should be independent of a specific course/enrollment. If we are worried about costs, we can set broad rules for when a user is allowed to verify:

The user is enrolled in a course run that requires verification AND either of the following is true:

1. The  user is currently unverified.
2. The user's verification is expiring in the next 7 days. (Again, this is broad. We could tighten this date to be relative to the earliest unexpired verification deadline of a user's enrollments.)

Aside: Personally, I'm starting to question the need for upgrade/verification deadlines; but, that is a separate conversation.

The URLs can be simplified: `/verify_student/start_flow/`.

mjfrey replied:
I understand, however we should do this optimization after this is moved from LMS.  The current receipt page shows this link for each purchase (course_run).

clintonb replied:
I don't necessarily consider this an optimization as much as it is paying down debt. Punting this now is just transferring the debt from one system to another.

At a minimum, the verification message should be displayed only once, and debt tickets should be created.

mjfrey replied:
I will create 2 tickets.  One for re-working the LMS verify flow as stated above and the other for this page/view to use the new endpoint.

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -113,3 +116,92 @@ def get_context_data(self, **kwargs):
+        return super(ReceiptResponseView, self).dispatch(*args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
+        context = self.get_context_data(**kwargs)
```
You've done a bit too much here, and made this slightly more complicated than it needs to be. I'll post a simpler version shortly.

clintonb replied:
I had an idea, which lead to more ideas, which lead to https://github.com/edx/ecommerce/commits/clintonb/receipt-temp. Feel free to incorporate these changes.

clintonb commented on ecommerce/extensions/api/serializers.py:
```diff
@@ -199,44 +201,79 @@ class Meta(object):
 
 class LineSerializer(serializers.ModelSerializer):
     """Serializer for parsing line item data."""
+    formatted_price = serializers.SerializerMethodField()
```
By which I mean in the template, not via JavaScript.

marjev replied:
touché

clintonb commented on ecommerce/extensions/checkout/views.py:
```diff
@@ -113,3 +116,92 @@ def get_context_data(self, **kwargs):
+    def dispatch(self, *args, **kwargs):
+        """
+        Customers should only be able to view their receipts when logged in. To avoid blocking responses
+        from payment processors which POST back to the page, the view must be CSRF-exempt.
```
Payment processors should no longer be posting to this page.

clintonb commented on ecommerce/templates/oscar/checkout/_credit_provider_data.html:
```diff
@@ -0,0 +1,38 @@
+      data-track-label="{{ provider.course_key }}"
+      data-track-type="click"
+      data-username="{{ name }}"
+      href="{{ provider.url }}">
```
This won't work as expected in all cases. Particularly, for providers with which we have a full integration (e.g. we send signed data), this will fail. *sigh* I have no good alternatives at the moment. Unfortunately, a resolution to this issue (due to signing key security concerns) may not come quickly.

FYI @mjfrey

clintonb replied:
Hmm...you could clean this up by forcing all credit requests to be initiated from the student dashboard rather than the receipt page. We already support that behavior. I'm not certain why we ever put this functionality on the receipt page.

marjev commented on ecommerce/extensions/payment/views.py:
```diff
@@ -272,6 +272,38 @@ def post(self, request):
+            )
+            return self.redirect_to_receipt_page_on_success(request, receipt_page_url)
+        else:
+            context = self.get_context_data()
```
I'll remove this from the PR and open a new one. I agree with your statement.

Instead of inheriting logic from `CybersourceNotifyView`, I'll move the logic that handles the Cybersource response. We can go from there in a new PR.

marjev commented:
@clintonb @mjfrey All the tests passed. Green coverage as well. I think I addressed all the comments. Please take another look.

I would like to keep this as simple as this and open a new PR for the responsive design once we're sure the current changes (text, design) on the Receipt Page is what we want.

marjev commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,166 @@ def test_post_returns_payment_support_email_in_context(self):
+        All other users should get the 404 status.
+        """
+        staff_user = self.create_user(is_staff=True)
+        other_user = self.create_user()
```
I would rather split this into two test case scenarios to avoid branching.

marjev replied:
Done.

clintonb commented on ecommerce/core/models.py:
```diff
@@ -485,6 +485,7 @@ def is_verified(self, site):
                     cache.set(cache_key, verification, cache_timeout)
             return verification
         except HttpNotFoundError:
+            log.info('No verification data found for [%s]', self.username)
```
Use `log.debug`to avoid littering Splunk.

clintonb commented on ecommerce/extensions/basket/tests/test_views.py:
```diff
@@ -368,7 +368,7 @@ def test_response_success(self, benefit_type, benefit_value):
         self.assertEqual(len(response.context['formset_lines_data']), 1)
         line_data = response.context['formset_lines_data'][0][1]
-        self.assertEqual(line_data['benefit_value'], format_benefit_value(benefit))
+        self.assertEqual(line_data['benefit_value'], format_benefit_value(benefit, settings.OSCAR_DEFAULT_CURRENCY))
```
`basket.currency`

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        )
+        return self.create_order(credit=credit)
+
+    def test_login_required_post_request(self):
```
The view should now be `GET` only, since the CyberSource code has been removed from the receipt page.

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        response = self.client.post(self.path)
+        self.assertEqual(response.status_code, 302)
+
+    def test_login_required_get_request(self):
```
Remove.

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        """ The view should redirect to the login page if the user is not logged in. """
+        self.client.logout()
+        response = self.client.post(self.path)
+        self.assertEqual(response.status_code, 302)
```
```python
testserver_login_url = self.get_full_url(reverse(settings.LOGIN_URL))
expected_url = '{path}?next={next}'.format(path=testserver_login_url, next=urllib.quote(self.path))
self.assertRedirects(response, expected_url, target_status_code=302)
```

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        order_number = 'ABC123'
+        response = self.client.get('{path}?order_number={order_number}'.format(
+            order_number=order_number,
+            path=self.path
```
Nit: Order the kwargs as they appear in the format string.

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+
+    def test_get_payment_method_payment_processor(self):
+        """
+        Payment Processor name should be displayed as the Payment method
```
We should not be displaying the payment processor name to consumers. "CyberSource" means nothing to them. Display the card/payment type.

marjev replied:
Here's the InVision design:
https://projects.invisionapp.com/share/7D76K432C#/screens/206952822_Receipt_Whole_Cart_Coupon
Looking at it, I came to a conclusion that we want to display Payment Processor name. 
What should be displayed in case payment processor was used?

clintonb replied:
PayPal is an exception since it is both a payment processor class and a payment type. You should be displaying payment/card type from the `Source` model.

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        when a Credit card was used to purchase a product.
+        """
+        order = self.create_order()
+        source = factories.SourceFactory(order=order)
```
`source = factories.SourceFactory(order=order, card_type='Dummy Card', label='Test')`

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        """
+        self.client.logout()
+        self.client.login(username=user.username, password=self.password)
+        return self.client.get('{path}?order_number={order_number}'.format(
```
This bit of code is duplicated numerous times. Create a helper method:

```python
def get_receipt_response(self, order):
  url = '{path}?order_number={order_number}'.format(path=self.path, order_number=order.number)
  return self.client.get(url)
```

clintonb commented on ecommerce/extensions/checkout/tests/test_views.py:
```diff
@@ -121,3 +145,178 @@ def test_post_returns_payment_support_email_in_context(self):
+        seat = order.lines.first().product
+        body = {'display_name': 'Hogwarts'}
+
+        httpretty.register_uri(
```
We are no longer calling the Credit API, so this should go.

clintonb commented on ecommerce/templates/edx/checkout/receipt.html:
```diff
@@ -0,0 +1,155 @@
+         data-currency="{{ order.currency }}"
+         data-fire-tracking-events="{{ fire_tracking_events|yesno:"true,false" }}"
+         data-order-id="{{ order.number }}"
+         data-total-amount="{{ order.total_excl_tax }}">
```
`order.total_incl_tax`

clintonb commented on ecommerce/templates/edx/checkout/receipt.html:
```diff
@@ -0,0 +1,155 @@
+  <div class="receipt">
+  <div class="container content-container">
+    <div id="receipt-container"
+         class="pay-and-verify"
```
Remove this class. This seems to be leftover from the move from LMS.

clintonb commented on ecommerce/templates/edx/checkout/receipt.html:
```diff
@@ -0,0 +1,155 @@
+{% endblock navbar %}
+
+{% block content %}
+  <div class="receipt">
```
There is a lot of nesting going on here. Can we combine these three levels of `<div>`?

clintonb commented on ecommerce/templates/edx/checkout/receipt.html:
```diff
@@ -0,0 +1,155 @@
+              </div>
+            </div>
+
+            {#            TODO: Make this table responsive.#}
```
This needs to be done.

marjev replied:
Let's do this in the separate PR. We'll have to make the entire page to be responsive.

Also, I want to make sure that texts and the "original" design is done properly before I focus on the responsive design.

clintonb replied:
If responsiveness or accessibility support are not happening on this PR, this PR cannot be merged to master. We just met our DOJ obligations regarding accessibility, and we aren't going to risk running afoul of them again. 25% of our users are on mobile, and we will not be alienating them.

clintonb commented on ecommerce/templates/edx/checkout/receipt_not_found.html:
```diff
@@ -0,0 +1,45 @@
+            <p>
+              {% blocktrans %}
+                The specified order could not be located.
+                Please ensure that the URL is correct, and try again.
```
Nit: this can go on one line.

clintonb commented on ecommerce/templates/oscar/checkout/_credit_provider_data.html:
```diff
@@ -0,0 +1,38 @@
+{% load i18n %}
```
This is not used, and should be deleted.

clintonb commented on ecommerce/templates/oscar/checkout/error.html:
```diff
@@ -22,6 +22,6 @@
             {% endblocktrans %}
         {% endwith %}</p>
-    
+
```
This empty line is not necessary.

clintonb commented on ecommerce/static/sass/partials/views/_receipt.scss:
```diff
@@ -0,0 +1,274 @@
+        text-decoration: line-through;
+    }
+
+    .order-amount {
```
Why is this necessary? Why not let the browser fit it all?

marjev replied:
I'll remove that. I don't have an explanation for most of these changes because I pulled them from Joyce's original PR. :/

clintonb commented on ecommerce/extensions/offer/utils.py:
```diff
@@ -46,12 +46,13 @@ def get_discount_value(discount_percentage, product_price):
 
 
-def format_benefit_value(benefit):
+def format_benefit_value(benefit, currency):
```
Are these changes still necessary?

clintonb commented on ecommerce/extensions/checkout/utils.py:
```diff
@@ -2,7 +2,7 @@
 from babel.numbers import format_currency
 from django.conf import settings
-from django.utils.translation import get_language, to_locale
+from django.utils.translation import get_language, to_locale, ugettext_lazy as _
```
Are any changes to this file still necessary?

clintonb commented:
That can be merged once it's ready. It is independent of the receipt page. 

marjev commented:
@clintonb Comments addressed.

Here are two screenshots showing desktop and mobile design, respectively.

![image](https://cloud.githubusercontent.com/assets/6833568/20719120/4c5fa01e-b65b-11e6-9194-75761ef3f221.png)

![image](https://cloud.githubusercontent.com/assets/6833568/20719143/5f72271c-b65b-11e6-930c-825c79176f51.png)

clintonb commented on ecommerce/core/templatetags/core_extras.py:
```diff
@@ -38,6 +41,34 @@ def do_captureas(parser, token):
+@register.filter(name='benefit_discount')
+def benefit_discount(benefit):
+    """
+    Format benefit value.
```
Format it how? Please provide an example in the docstring.

clintonb commented on ecommerce/core/templatetags/core_extras.py:
```diff
@@ -38,6 +41,34 @@ def do_captureas(parser, token):
     return CaptureasNode(nodelist, args)
 
 
+@register.filter(name='organization')
```
course_organization

clintonb commented on ecommerce/core/templatetags/core_extras.py:
```diff
@@ -38,6 +41,34 @@ def do_captureas(parser, token):
 
 
+@register.filter(name='organization')
+def course_organization(course_key):
```
This is not used anywhere. It, and its tests, should be deleted.

clintonb commented on ecommerce/core/templatetags/core_extras.py:
```diff
@@ -38,6 +41,34 @@ def do_captureas(parser, token):
+    Returns:
+        str: String value containing formatted benefit value and type.
+    """
+    return format_benefit_value(benefit)
```
This method is implemented incorrectly for fixed price discounts. It will always return a string formatted for USD.  This will cause issues for our open source users who do not use USD. This isn't a show-stopper for this PR; but, please file a ticket to fix this, @mjfrey.

marjev replied:
@mjfrey I think the reason behind this implementation is the locale error that you told me about. It would be great if the ticket has a concrete description of the problem.

clintonb commented on ecommerce/core/templatetags/core_extras.py:
```diff
@@ -38,6 +41,34 @@ def do_captureas(parser, token):
+
+
+@register.filter(name='benefit_discount')
+def benefit_discount(benefit):
```
This doesn't belong in `core`. `core` is intended for components that will be used throughout the entire project. Move this tag to an `offer_extras` or `offer_tags` in the `offer` app.

marjev commented:
@clintonb I apologize for pinging you before, without checking what I've pushed. Some of the changes were stashed locally and I forgot to apply those before.

I think all the comments are addressed now.

Latest screenshots:
<img src="https://cloud.githubusercontent.com/assets/6833568/20794390/56a4e166-b7cc-11e6-97b9-ee2fa4d3dfda.png" alt="Desktop version" width="75%" height="auto">

<img src="https://cloud.githubusercontent.com/assets/6833568/20794517/0078f47a-b7cd-11e6-8e2b-15c95403ffe5.png" alt="Mobile version" width="50%" height="50%">

marjev commented:
@clintonb anything else needed on this PR?
